capsules/src/virtual_uart.rs

```diff
@@ -0,0 +1,217 @@
+//! let console = static_init!(
+//!     capsules::console::Console<UartDevice>,
+//!     capsules::console::Console::new(
+//!         console_uart,
+//!         115200,
+//!         &mut capsules::console::WRITE_BUF,
+//!         &mut capsules::console::READ_BUF,
+//!         kernel::Grant::create()
+//!     )
+//! );
+//! hil::uart::UART::set_client(console_uart, console);
+//! ```
+
+use core::cell::Cell;
+use core::cmp;
+
+use kernel::common::cells::{OptionalCell, TakeCell};
+use kernel::common::{List, ListLink, ListNode};
+use kernel::hil;
+
+pub static mut RX_BUF: [u8; 64] = [0; 64];
+
+pub struct UartMux<'a> {
+    uart: &'a hil::uart::UART,
+    devices: List<'a, UartDevice<'a>>,
+    receive_in_progress: Cell<bool>,
+    inflight: OptionalCell<&'a UartDevice<'a>>,
+    rx_buffer: TakeCell<'static, [u8]>,
+}
+
+impl<'a> hil::uart::Client for UartMux<'a> {
+    fn transmit_complete(&self, tx_buffer: &'static mut [u8], error: hil::uart::Error) {
+        self.inflight.map(move |device| {
+            self.inflight.clear();
+            device.transmit_complete(tx_buffer, error);
+        });
+        self.do_next_op();
+    }
+
+    fn receive_complete(
+        &self,
+        rx_buffer: &'static mut [u8],
+        rx_len: usize,
+        error: hil::uart::Error,
+    ) {
+        // Pass the data to each receiver that wants to receive messages.
+        self.devices.iter().for_each(|device| {
+            if device.receiver {
+                device.rx_buffer.take().map(|rxbuf| {
+                    let len = cmp::min(rx_len, device.rx_len.get());
+                    for i in 0..len {
+                        rxbuf[i] = rx_buffer[i];
+                    }
+                    device.receive_complete(rxbuf, len, error);
+                });
+            }
+        });
+        self.rx_buffer.replace(rx_buffer);
+        self.receive_in_progress.set(false);
+    }
+}
+
+impl<'a> UartMux<'a> {
+    pub fn new(uart: &'a hil::uart::UART, rx_buffer: &'static mut [u8]) -> UartMux<'a> {
+        UartMux {
+            uart: uart,
+            devices: List::new(),
+            receive_in_progress: Cell::new(false),
+            inflight: OptionalCell::empty(),
+            rx_buffer: TakeCell::new(rx_buffer),
+        }
+    }
+
+    fn do_next_op(&self) {
+        if self.inflight.is_none() {
+            let mnode = self.devices.iter().find(|node| node.operation.is_some());
+            mnode.map(|node| {
+                node.tx_buffer.take().map(|buf| {
+                    node.operation.map(move |op| match op {
+                        Operation::Transmit { len } => self.uart.transmit(buf, *len),
+                    });
+                });
+                node.operation.clear();
+                self.inflight.set(node);
+            });
+        }
+    }
+
+    /// If we are not currently listening, start listening.
+    fn start_receive(&self, rx_len: usize) {
+        if self.receive_in_progress.get() == false {
```

Comment:
Not that it hurts anything to have, but is `receive_in_progress` actually doing anything? I think if you just eliminated the flag completely nothing would functionally change (i.e. here the takecell would be empty if one was already in flight and skip over the map, above once replaced it'd be ready to go again)